### ISSUE TITLE
Panoptic Results

### ISSUE THREAD

**Issue by shwoo1219** [LaoYang1994/SOGNet-Dev#1]
Many thanks for sharing the detectron2 based code of SOGNet.
I trained the UPSNet (SOGNet without relation loss) and got the following panoptic results.

\|        \|   PQ   \|   SQ   \|   RQ   \|  #categories  \|
--
\|:------:\|:------:\|:------:\|:------:\|:-------------:\|
\|  All   \| 43.334 \| 80.465 \| 52.806 \|      133      \|
\| Things \| 31.357 \| 77.841 \| 39.840 \|      80       \|
\| Stuff  \| 61.411 \| 84.426 \| 72.378 \|      53       \|

As noted, the code seems to have some bugs (* compared to the official score, PQ_things is too low whereas PQ_stuff is too high).
In your opinion, where is the most suspicious part of the code? 
Maybe I can check that part quickly.

**Comment by LaoYang1994** [comment]
@shwoo93 Hi, thanks for your attention. Actually, I haven't found where the bug is. But there is one thing you can do.
1. Merge the instance results and semantic results using heuristic method like panoptic-fpn, which I have added in this code. If the performance looks reasonable, we can think there are some bugs during the inference. If so, you can compare the inference part in d2-based code and upsnet-based code.
2. I reimplement the code in detectron2 just for effiency, memory saving and multi-imgs per gpu support. There may be some bugs in this process. 
3. Ground Truth. UPS-code prepares the pan-gt in an online way, which may affect running efficiency. So I prepare the gt in an offline way. There may be some bugs in this process.

I think the first is most likely to have problems.

Besides, I think you have used a wrong config. Line https://github.com/LaoYang1994/SOGNet-Dev/blob/250d9ccddcd9f92ede87b5424ef9396f6f22f562/configs/sognet_R_50_1x.yaml#L63 is just for debug and you should use https://github.com/LaoYang1994/SOGNet-Dev/blob/250d9ccddcd9f92ede87b5424ef9396f6f22f562/configs/sognet_R_50_1x.yaml#L62. You will find the $PQ_st$ is normal.

If you have any question, please let me know. I will try to find the problem after my graduation.

**Comment by shwoo1219** [comment]
Thanks for your kind reply.

I have trained again with the correct config file and got the following results

heuristic combine

\|  All   \| 39.451 \| 78.245 \| 48.114 \|      133      \|
\| Things \| 43.917 \| 80.785 \| 53.010 \|      80       \|
\| Stuff  \| 32.711 \| 74.410 \| 40.724 \|      53       \|

panoptic head

\|  All   \| 27.482 \| 72.880 \| 36.348 \|      133      \|
\| Things \| 27.425 \| 72.414 \| 36.787 \|      80       \|
\| Stuff  \| 27.570 \| 73.584 \| 35.684 \|      53       \|

As you mentioned, the heuristic combine seems to produce reasonable results while the panoptic head fails to. I have checked the inference code of panoptic head.py and pan_seg_postprocess.py, but could not find any critical issues there (I only found some implementation details are different but seem to not affect the general operation).

For the GT part, I see the difference between d2-based code and the original code lies in handling the inconsistency between instance annotation and panoptic annotation. In the d2-based code, I see lots of gt boxes are not contributing to the final panoptic logit but ignored during training (i.e., if the pan_id of gt mask is to be -1, its region in the final panoptic logit is ignored). Will this affect the performance? Not that sure though.

**Comment by LaoYang1994** [comment]
@shwoo93 Hi， I‘m sorry for my late reply because of my graduation issues. I will spend some time checking the code.
